packages/gatsby/src/query/query-compiler.js

```diff
@@ -1,384 +1,400 @@
+        operations.push(def)
+      } else if (def.kind === Kind.FRAGMENT_DEFINITION) {
+        if (definitionsByName.has(name)) {
+          const otherDef = definitionsByName.get(name)
+          if (text !== otherDef.text) {
+            addError(
+              duplicateFragmentError({
+                name,
+                leftDefinition: {
+                  def,
+                  filePath,
+                  text,
+                },
+                rightDefinition: otherDef,
+              })
+            )
+            definitionsByName.delete(name)
+          }
+          return
+        }
+      }
+      definitionsByName.set(name, {
+        def,
+        filePath,
+        text,
+        isFragment: def.kind === Kind.FRAGMENT_DEFINITION,
+      })
+    })
+  }
+  return {
+    definitionsByName,
+    operations,
+  }
+}
+
+const processDefinitions = ({
+  schema,
+  operations,
+  definitionsByName,
+  addError,
+  parentSpan,
+}) => {
+  const processedQueries: Queries = new Map()
+
+  const usedFragmentsForFragment = new Map()
+
+  const fragmentNames = Array.from(definitionsByName.entries())
+    .filter(([_, def]) => def.isFragment)
+    .map(([name, _]) => name)
+
+  for (const operation of operations) {
+    const name = operation.name.value
+    const originalDefinition = definitionsByName.get(name)
+    const filePath = definitionsByName.get(name).filePath
+    if (processedQueries.has(filePath)) {
+      const otherQuery = processedQueries.get(filePath)
+
+      addError(
+        multipleRootQueriesError(
+          filePath,
+          originalDefinition.def,
+          otherQuery && definitionsByName.get(otherQuery.name).def
         )
       )
-    } catch (error) {
-      const { formattedMessage, docName, message, codeBlock } = graphqlError(
-        namePathMap,
-        nameDefMap,
-        error
-      )
-      nameErrorMap.set(docName, { formattedMessage, message, codeBlock })
-      boundActionCreators.queryExtractionGraphQLError({
-        componentPath: namePathMap.get(docName),
-        error: formattedMessage,
-      })
 
-      const filePath = namePathMap.get(docName)
-      addError(errorParser({ message, filePath }))
+      actions.queryExtractionGraphQLError({
+        componentPath: filePath,
+      })
+      continue
+    }
 
-      return false
+    const usedFragments = new Set()
+    const stack = [operation]
+    let missingFragment = false
+    while (stack.length > 0) {
+      const def = stack.pop(operation)
+      visit(def, {
+        [Kind.FRAGMENT_SPREAD]: node => {
+          const name = node.name.value
+          if (usedFragmentsForFragment.has(name)) {
+            usedFragmentsForFragment.get(name).forEach(derivedFragmentName => {
+              usedFragments.add(derivedFragmentName)
+            })
+            usedFragments.add(name)
+          } else if (definitionsByName.has(name)) {
+            stack.push(definitionsByName.get(name).def)
```

Comment:
add a comment why we need to do this? I guess it's to make sure we catch nested fragments.